custom_components/unifiprotect/services.py

```diff
@@ -22,65 +24,101 @@ def _async_all_ufp_instances(hass: HomeAssistant) -> list[ProtectApiClient]:
-    device_registry = dr.async_get(hass)
 
-    instances: list[tuple[str, ProtectApiClient]] = []
+    instances: list[tuple[dr.DeviceEntry, ProtectApiClient]] = []
     for device_id in referenced.referenced_devices:
-        if not (device_entry := device_registry.async_get(device_id)):
-            raise HomeAssistantError(f"No device found for device id: {device_id}")
-
-        name = str(device_entry.name_by_user or device_entry.name)
-        macs = [
-            # MAC addresses in UFP are always caps
-            cval.replace(":", "").upper()
-            for ctype, cval in device_entry.connections
-            if ctype == dr.CONNECTION_NETWORK_MAC
-        ]
-        ufp_instances = [
-            (name, i)
-            for i in _async_all_ufp_instances(hass)
-            if i.bootstrap.nvr.mac in macs
-        ]
-        if not ufp_instances:
-            raise HomeAssistantError(
-                f"No UniFi Protect NVR found for device ID: {device_id}"
-            )
-        instances += ufp_instances
+        instances.append(_async_get_ufp_instances(hass, device_id))
 
     return instances
 
 
 async def add_doorbell_text(hass: HomeAssistant, call: ServiceCall) -> None:
     message: str = call.data[CONF_MESSAGE]
     instances = _async_get_protect_from_call(hass, call)
-    await asyncio.gather(
-        *(i.bootstrap.nvr.add_custom_doorbell_message(message) for _, i in instances)
-    )
+
+    try:
+        await asyncio.gather(
+            *(
+                i.bootstrap.nvr.add_custom_doorbell_message(message)
+                for _, i in instances
+            )
+        )
+    except (BadRequest, ValidationError) as err:
+        raise HomeAssistantError(str(err)) from err
 
 
 async def remove_doorbell_text(hass: HomeAssistant, call: ServiceCall) -> None:
     message: str = call.data[CONF_MESSAGE]
     instances = _async_get_protect_from_call(hass, call)
-    await asyncio.gather(
-        *(i.bootstrap.nvr.remove_custom_doorbell_message(message) for _, i in instances)
-    )
+
+    try:
+        await asyncio.gather(
+            *(
+                i.bootstrap.nvr.remove_custom_doorbell_message(message)
+                for _, i in instances
+            )
+        )
+    except (BadRequest, ValidationError) as err:
+        raise HomeAssistantError(str(err)) from err
 
 
 async def set_default_doorbell_text(hass: HomeAssistant, call: ServiceCall) -> None:
     message: str = call.data[CONF_MESSAGE]
     instances = _async_get_protect_from_call(hass, call)
-    await asyncio.gather(
-        *(i.bootstrap.nvr.set_default_doorbell_message(message) for _, i in instances)
-    )
+
+    try:
+        await asyncio.gather(
+            *(
+                i.bootstrap.nvr.set_default_doorbell_message(message)
+                for _, i in instances
+            )
+        )
+    except (BadRequest, ValidationError) as err:
+        raise HomeAssistantError(str(err)) from err
```

Comment:
Similar code to above.  Can we make it its own function and pass in the coro to call?